src/map.cpp

```diff
@@ -605,69 +605,141 @@ bool map::vehicle_falling( vehicle &veh )
     return true;
 }
 
-float map::vehicle_traction( vehicle &veh ) const
+float map::traction_mult( const tripoint &p ) const
 {
-    const tripoint pt = veh.global_pos3();
-    // TODO: Remove this and allow amphibious vehicles
-    if( !veh.floating.empty() ) {
-        return vehicle_buoyancy( veh );
+    const ter_t ter = ter_at( p );
```

Comment:
Please don't copy the terrain data, a reference should work just fine. Same in the next function.